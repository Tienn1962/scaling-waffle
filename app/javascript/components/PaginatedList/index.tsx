import { ArrowLeftIcon, ArrowRightIcon } from "@heroicons/react/outline";
import React from "react";
import ReactPaginate from "react-paginate";

interface Props<T> {
  data: T[];
  count: number;
  limit: number;
  element: (datum: T) => React.ReactNode;
  onPageChange: (page: number) => void;
}

export default <T,>(props: Props<T>) => {
  const handlePageChange = (selectedItem: { selected: number }) => {
    props.onPageChange(selectedItem.selected);
  };

  return (
    <>
      <div className="flex flex-row flex-wrap justify-center">
        {props.data.map((datum) => props.element(datum))}
      </div>
      <div className="flex flex-row justify-center">
        <ReactPaginate
          className="join fixed bottom-0"
          activeLinkClassName="bg-black"
          pageLinkClassName="join-item btn btn-square no-animation"
          previousClassName="join-item btn btn-square no-animation"
          nextClassName="join-item btn btn-square no-animation"
          breakClassName="join-item btn btn-square no-animation"
          pageCount={Math.ceil(props.count / props.limit)}
          pageRangeDisplayed={5}
          onPageChange={handlePageChange}
          nextLabel={<ArrowRightIcon className="h-6 w-6" />}
          previousLabel={<ArrowLeftIcon className="h-6 w-6" />}
        />
      </div>
    </>
  );
};
