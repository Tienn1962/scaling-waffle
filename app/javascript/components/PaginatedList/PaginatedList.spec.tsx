import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import PaginatedList from ".";

it("renders a list of items with a template", () => {
  const component = render(
    <PaginatedList
      data={["Hello world!", "Goodbye world!"]}
      count={2}
      limit={50}
      item={(datum) => <div key={datum}>{datum}</div>}
      onPageChange={() => {}}
    />
  );

  expect(component).toMatchSnapshot();
});

it("emits an event on page change", async () => {
  const callback = jest.fn();

  render(
    <PaginatedList
      data={["Hello world!", "Goodbye world!"]}
      count={2}
      limit={1}
      item={(datum) => <div key={datum}>{datum}</div>}
      onPageChange={(page) => {
        callback();
      }}
    />
  );

  const secondPageLink = screen.getByText("2");
  fireEvent.click(secondPageLink);
  expect(callback).toHaveBeenCalled();
});
