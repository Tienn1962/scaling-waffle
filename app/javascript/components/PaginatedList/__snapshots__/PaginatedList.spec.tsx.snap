// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of items with a template 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-row flex-wrap justify-center"
      >
        <div>
          Hello world!
        </div>
        <div>
          Goodbye world!
        </div>
      </div>
      <div
        class="flex flex-row justify-center"
      >
        <ul
          aria-label="Pagination"
          class="join fixed bottom-0"
          role="navigation"
        >
          <li
            class="join-item btn btn-square no-animation disabled"
          >
            <a
              aria-disabled="true"
              aria-label="Previous page"
              class=" "
              rel="prev"
              role="button"
              tabindex="-1"
            >
              <svg
                aria-hidden="true"
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 19l-7-7m0 0l7-7m-7 7h18"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </a>
          </li>
          <li
            class="selected"
          >
            <a
              aria-current="page"
              aria-label="Page 1 is your current page"
              class="join-item btn btn-square no-animation bg-black"
              rel="canonical"
              role="button"
              tabindex="-1"
            >
              1
            </a>
          </li>
          <li
            class="join-item btn btn-square no-animation disabled"
          >
            <a
              aria-disabled="true"
              aria-label="Next page"
              class=" "
              rel="next"
              role="button"
              tabindex="-1"
            >
              <svg
                aria-hidden="true"
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14 5l7 7m0 0l-7 7m7-7H3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-row flex-wrap justify-center"
    >
      <div>
        Hello world!
      </div>
      <div>
        Goodbye world!
      </div>
    </div>
    <div
      class="flex flex-row justify-center"
    >
      <ul
        aria-label="Pagination"
        class="join fixed bottom-0"
        role="navigation"
      >
        <li
          class="join-item btn btn-square no-animation disabled"
        >
          <a
            aria-disabled="true"
            aria-label="Previous page"
            class=" "
            rel="prev"
            role="button"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 19l-7-7m0 0l7-7m-7 7h18"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </a>
        </li>
        <li
          class="selected"
        >
          <a
            aria-current="page"
            aria-label="Page 1 is your current page"
            class="join-item btn btn-square no-animation bg-black"
            rel="canonical"
            role="button"
            tabindex="-1"
          >
            1
          </a>
        </li>
        <li
          class="join-item btn btn-square no-animation disabled"
        >
          <a
            aria-disabled="true"
            aria-label="Next page"
            class=" "
            rel="next"
            role="button"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M14 5l7 7m0 0l-7 7m7-7H3"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
